### ISSUE TITLE
Semantics of MAP in probabilistic programming

### ISSUE THREAD

**Issue by motownblue** [p2t2/figaro#603]
In ordinary probabilistic models with a fixed set of variables, this is easy to understand. But when you have different sets of variables in different branches of the program, it can be hard to understand. For example, here’s a program:

x = flip(0.2)
y = if x Normal(0,1) else { z1 = Normal(1, 1), z2 = Normal(1, 1), z1 + z2 }

**Comment by wkretschmer** [comment]
For a while I thought that this type of measure-theoretic problem only applied to MAP inference. However, I've found inconsistencies elsewhere. For example:
```
val x = Flip(0.4)
val y = If(x, Constant(0.0), Uniform(-0.25, 0.25))
y.observe(0.0)

println(Importance.probability(x, true))
println(MetropolisHastings.probability(x, true))
```
This outputs:
```
0.25054711436253346
1.0
```
At some point, we added the ability of likelihood weighting to handle hard observations on continuous elements by looking at the densities of the observed values. So, Importance runs into trouble here because it tries to compare mass and density functions for the two result elements. On the other hand, MH produces the "correct" answer because it rejects any proposal where x=false.

This type of problem has appeared in other probabilistic programming contexts, but no one else has a complete solution. For example, see the Anglican developers' discussion of the "[Indian GPA problem](http://www.robots.ox.ac.uk/~fwood/anglican/examples/viewer/?worksheet=indian-gpa)", a similar mixed discrete/continuous model.

**Comment by apfeffer** [comment]
Yes, it’s an interesting research problem to characterize and develop inference algorithms that are correct for these types of problems. Personally, I favor the draconian solution of disallowing point observations on continuous variables.

**Comment by bruttenberg** [comment]
FYI, the Indian GPA problem is from Stuart’s unpublished paper that  on this topic (the one we asked about).

+  +  +  +  +  +  +  +  +  +  +  +  +  +
Brian Ruttenberg, PhD
Senior Scientist
Decision Management Systems Division
Government Services
Charles River Analytics Inc.
617.491.3474 x730
www.cra.com<http://www.cra.com/>

**Comment by wkretschmer** [comment]
Here's one possibility for a working definition of MAP in probabilistic programs. A common use case among models with uncertain structure is to have some variable that represents the number of other variables present. For example, in pseudocode:
```
x = Poisson(2)
for(i <- 1 to x) y[i] = Normal(i, 1)
z = y.sum
// some soft evidence on z
```
A user calling for a MAP query on this model probably wants the most likely values of the array y. Strictly speaking, the most likely assignment is the empty array because the dimension of the joint density function depends on x, and assignments with lower dimension are infinitely more likely. This isn't a meaningful answer: in reality, what the user probably wants to do is (1) find the marginal MAP value of x, then (2) find the most likely values of y given this assignment to x.

Let's put this in slightly more general Figaro terms. For a model consisting of two permanent elements (a parent element and a Chain depending on the parent), we might define the MAP assignment to consist of the marginal MAP value of the parent, coupled with the conditional MAP value of the Chain given the value of the parent. We can compute this in two steps. It's easy to extend this definition when we have additional permanent elements. If those elements don't depend on the Chain, we maximize them jointly with the parent in the first step. Otherwise, we maximize them jointly with the Chain in the second step. With a bit more work, we can also extend this to models with multiple Chains, including nested Chains in result elements. We just have to have MAP values for a Chain's parent before performing MAP on the Chain and elements that depend on the Chain.

Another way of thinking about this definition is that we build up the MAP assignment by repeatedly performing marginal MAP on parts of the model where structure is certain, given the MAP values found so far. Once we have a MAP value on the parent of a Chain, the result element of the Chain is no longer uncertain, and so we can compute its MAP value.

The only problem I see with this definition is that computing MAP this way will be impractical for most models, including some where a standard MAP algorithm would work fine. Can we exploit the structure of the marginal MAP subproblems? Notably, each subproblem only involves MAPping elements that are constrained to be top-level.

**Comment by apfeffer** [comment]
This is interesting. We could define MAP to mean marginal MAP for the top level and conditional MAP for the lower levels. Is that what you’re suggesting? It would be easy to implement in SFI.

Avi
++++++++++++++++++++++
Avi Pfeffer
Chief Scientist
Charles River Analytics, Inc.
(617) 491-3474x513
apfeffer@cra.com<mailto:apfeffer@cra.com>
www.cra.com<http://www.cra.com>

On Jul 28, 2016, at 1:19 PM, William Kretschmer <notifications@github.com<mailto:notifications@github.com>> wrote:


Here's one possibility for a working definition of MAP in probabilistic programs. A common use case among models with uncertain structure is to have some variable that represents the number of other variables present. For example, in pseudocode:

x = Poisson(2)
for(i <- 1 to x) y[i] = Normal(i, 1)
z = y.sum
// some soft evidence on z


A user calling for a MAP query on this model probably wants the most likely values of the array y. Strictly speaking, the most likely assignment is the empty array because the dimension of the joint density function depends on x, and assignments with lower dimension are infinitely more likely. This isn't a meaningful answer: in reality, what the user probably wants to do is (1) find the marginal MAP value of x, then (2) find the most likely values of y given this assignment to x.

Let's put this in slightly more general Figaro terms. For a model consisting of two permanent elements (a parent element and a Chain depending on the parent), we might define the MAP assignment to consist of the marginal MAP value of the parent, coupled with the conditional MAP value of the Chain given the value of the parent. We can compute this in two steps. It's easy to extend this definition when we have additional permanent elements. If those elements don't depend on the Chain, we maximize them jointly with the parent in the first step. Otherwise, we maximize them jointly with the Chain in the second step. With a bit more work, we can also extend this to models with multiple Chains, including nested Chains in result elements. We just have to have MAP values for a Chain's parent before performing MAP on the Chain and elements that depend on the Chain.

Another way of thinking about this definition is that we build up the MAP assignment by repeatedly performing marginal MAP on parts of the model where structure is certain, given the MAP values found so far. Once we have a MAP value on the parent of a Chain, the result element of the Chain is no longer uncertain, and so we can compute its MAP value.

The only problem I see with this definition is that computing MAP this way will be impractical for most models, including some where a standard MAP algorithm would work fine. Can we exploit the structure of the marginal MAP subproblems? Notably, each subproblem only involves MAPping elements that are constrained to be top-level.
